panic with `ALTER TABLE t.test ADD a TIMESTAMP UNIQUE DEFAULT current_timestamp()`

```
panic: only column data should be modified, but the rowUpdater is configured otherwise
```

This is happening because the rowUpdater used by the schema change backfill mechanism is configured to update an index.

--- a/sql/backfill.go
+++ b/sql/backfill.go
@@ -268,7 +268,9 @@ func (sc *SchemaChanger) truncateAndBackfillColumnsChunk(
 		updateCols := append(added, dropped...)
 		// TODO(dan): Tighten up the bound on the requestedCols parameter to
 		// makeRowUpdater.
-		ru, err := makeRowUpdater(txn, tableDesc, updateCols, tableDesc.Columns)
+		ru, err := makeRowUpdater(
+			txn, tableDesc, updateCols, tableDesc.Columns, rowUpdaterOnlyColumns,
+		)
 		if err != nil {
 			return err
 		}

--- a/sql/rowwriter.go
+++ b/sql/rowwriter.go
@@ -270,6 +270,13 @@ type rowUpdater struct {
 	indexEntriesBuf []sqlbase.IndexEntry
 }
 
+type rowUpdaterType int
+
+const (
+	rowUpdaterDefault     rowUpdaterType = 0
+	rowUpdaterOnlyColumns rowUpdaterType = 1
+)
+
 // makeRowUpdater creates a rowUpdater for the given table.
 //
 // updateCols are the columns being updated and correspond to the updateValues
@@ -283,6 +290,7 @@ func makeRowUpdater(
 	tableDesc *sqlbase.TableDescriptor,
 	updateCols []sqlbase.ColumnDescriptor,
 	requestedCols []sqlbase.ColumnDescriptor,
+	updateType rowUpdaterType,
 ) (rowUpdater, error) {
 	updateColIDtoRowIndex := colIDtoRowIndexFromCols(updateCols)
 
@@ -301,6 +309,10 @@ func makeRowUpdater(
 
 	// Secondary indexes needing updating.
 	needsUpdate := func(index sqlbase.IndexDescriptor) bool {
+		if updateType == rowUpdaterOnlyColumns {
+			// Only update columns.
+			return false
+		}
 		// If the primary key changed, we need to update all of them.
 		if primaryKeyColChange {
 			return true

--- a/sql/tablewriter.go
+++ b/sql/tablewriter.go
@@ -260,7 +260,9 @@ func (tu *tableUpserter) init(txn *client.Txn) error {
 	if len(tu.updateCols) == 0 {
 		tu.fetchColIDtoRowIndex = colIDtoRowIndexFromCols(requestedCols)
 	} else {
-		tu.ru, err = makeRowUpdater(txn, tu.tableDesc, tu.updateCols, requestedCols)
+		tu.ru, err = makeRowUpdater(
+			txn, tu.tableDesc, tu.updateCols, requestedCols, rowUpdaterDefault,
+		)
 		if err != nil {
 			return err
 		}

--- a/sql/testdata/alter_table
+++ b/sql/testdata/alter_table
@@ -190,6 +190,14 @@ ALTER TABLE tt ADD COLUMN q DECIMAL NOT NULL
 statement ok
 ALTER table tt ADD COLUMN r DECIMAL
 
+# Ensure that a UNIQUE NOT NULL COLUMN can be added when there is no data in
+# the table.
+statement ok
+ALTER TABLE tt ADD COLUMN s DECIMAL UNIQUE NOT NULL
+
+statement ok
+ALTER TABLE tt ADD t DECIMAL UNIQUE DEFAULT 4.0
+
 query TTBT colnames
 SHOW COLUMNS FROM tt
 ----
@@ -197,6 +205,8 @@ Field  Type     Null   Default
 a      INT      false  NULL
 q      DECIMAL  false  NULL
 r      DECIMAL  true   NULL
+s      DECIMAL  false  NULL
+t      DECIMAL  true   4.0
 
 # Default values can be added and changed after table creation.
 statement ok
@@ -285,6 +295,11 @@ SELECT a,b FROM add_default WHERE e > d AND e - d < interval '10s'
 2 42
 3 10
 5 NULL
+
+# Adding a unique column to an existing table with data with a default value
+# is illegal
+statement error duplicate key value .* violates unique constraint \"add_default_f_key\"
+ALTER TABLE add_default ADD f TIMESTAMP UNIQUE DEFAULT current_timestamp()
 
 # Multiple columns can be added at once with heterogeneous DEFAULT usage
 statement ok

--- a/sql/update.go
+++ b/sql/update.go
@@ -163,7 +163,7 @@ func (p *planner) Update(n *parser.Update, desiredTypes []parser.Datum, autoComm
 		requestedCols = en.tableDesc.Columns
 	}
 
-	ru, err := makeRowUpdater(p.txn, en.tableDesc, updateCols, requestedCols)
+	ru, err := makeRowUpdater(p.txn, en.tableDesc, updateCols, requestedCols, rowUpdaterDefault)
 	if err != nil {
 		return nil, err
 	}

